nomad/structs/csi.go

```diff
@@ -874,6 +890,118 @@ func (p *CSIPlugin) DeleteAlloc(allocID, nodeID string) error {
 	return nil
 }
 
+// AddJob adds a job to the plugin and increments expected
+func (p *CSIPlugin) AddJob(job *Job, summary *JobSummary) {
+	p.UpdateExpectedWithJob(job, summary, false)
+}
+
+// DeleteJob removes the job from the plugin and decrements expected
+func (p *CSIPlugin) DeleteJob(job *Job, summary *JobSummary) {
+	p.UpdateExpectedWithJob(job, summary, true)
+}
+
+// UpdateExpectedWithJob maintains the expected instance count
+// we use the summary to add non-allocation expected counts
+func (p *CSIPlugin) UpdateExpectedWithJob(job *Job, summary *JobSummary, terminal bool) {
+	if summary == nil {
+		return
+	}
+
+	for _, tg := range job.TaskGroups {
+		count := tg.Count
+		if job.Type == JobTypeSystem {
+			s, ok := summary.Summary[tg.Name]
+			if !ok {
+				continue
+			}
+
+			count = s.Running + s.Queued + s.Starting
+		}
+
+		for _, t := range tg.Tasks {
+			if t.CSIPluginConfig == nil ||
+				t.CSIPluginConfig.ID != p.ID {
+				continue
+			}
+
+			// Change the correct plugin expected, monolith should change both
+			if t.CSIPluginConfig.Type == CSIPluginTypeController ||
+				t.CSIPluginConfig.Type == CSIPluginTypeMonolith {
+				if terminal {
+					p.ControllerJobs.Delete(job)
+				} else {
+					p.ControllerJobs.Add(job, count)
+				}
+			}
+
+			if t.CSIPluginConfig.Type == CSIPluginTypeNode ||
+				t.CSIPluginConfig.Type == CSIPluginTypeMonolith {
+				if terminal {
+					p.NodeJobs.Delete(job)
+				} else {
+					p.NodeJobs.Add(job, count)
+				}
+			}
+		}
+	}
+
+	p.ControllersExpected = p.ControllerJobs.Count()
+	p.NodesExpected = p.NodeJobs.Count()
+}
+
+type JobDescription struct {
+	Namespace string
+	ID        string
+	Version   uint64
+	Expected  int
+}
+
+type JobNamespacedDescriptions map[string]JobDescription
+type JobDescriptions map[string]JobNamespacedDescriptions
```

Comment:
Can we annotate the key for these maps? You can figure it out from context but having a comment here makes it easier to know what that `string` is at a glance.

Reply:
👍